pdr_backend/lake/alt_bar.py

```diff
@@ -0,0 +1,158 @@
+import logging
+from typing import List, Union
+import numpy as np
+
+from enforce_typing import enforce_types
+import pandas as pd
+
+# add config pair: volume_threshold in ppss
+
+
+@enforce_types
+def _extract_bars(data: pd.DataFrame, metric: str, threshold: float = 50000):
+    """
+    For loop which compiles the various bars: dollar, volume, or tick.
+
+    We did investigate the use of trying to solve this in a vectorised manner but found that a For loop worked well.
+
+    :param data: raw ohlcv data from kaiko Contains 6 columns - timestamp, open, high, low, close and volume.
+    :param metric: cum_ticks, cum_dollar_value, cum_volume
+    :param threshold: A cumulative value above this threshold triggers a sample to be taken.
+    :return: The financial data structure with the updated start timestamp since last bar.
+    """
+
+    list_bars = []
+    cache: list = []
+    start_tm = None
+    cum_ticks, cum_dollar_value, cum_volume, cache, high_price, low_price = (
+        0,
+        0.0,
+        0.0,
+        [],
+        -np.inf,
+        np.inf,
+    )
+
+    # Iterate over rows
+    for row in data.values:
+        # Set variables
+        timestamp = row[0]
+        open_price = float(row[1])
+        high = float(row[2])
+        low = float(row[3])
+        close = float(row[4])
+        volume = float(row[5])
+
+        # Calculations
+        cum_ticks += 1
+        dollar_value = close * volume
+        cum_dollar_value = cum_dollar_value + dollar_value
+        cum_volume = cum_volume + volume
+
+        # Check min max
+        if high > high_price:
+            high_price = high
+        if low <= low_price:
+            low_price = low
+
+        # Update cache
+        cache.append(
+            [
+                timestamp,
+                open_price,
+                high_price,
+                low_price,
+                close,
+                cum_volume,
+                cum_dollar_value,
+                cum_ticks,
+            ]
+        )
+
+        # If threshold reached then take a sample
+        if eval(metric) >= threshold:  # pylint: disable=eval-used
```

Comment:
at the beginning of this file, please do some checks on the "metric" string. If we are turning off the security feature of using `eval()` we should cover some bases to compensate